### ISSUE TITLE
InvalidXMLResponse("Couldn't find home-set.",)

### ISSUE THREAD

**Issue by whitedwarfsun** [pimutils/vdirsyncer#543]
I cannot get vdirsyncer to discover and sync my addressbook. Calendar and Tasks works fine.

The server is a Zarafa 7.1 setup (with zarafa-iCal) on Fedora 21 (I want to sync and migrate to another system).

My config:
```
[pair wdcontacts]
a = "wdconloc"
b = "wdconold"
collections = ["from a", "from b"]
# collections = null

[storage wdconloc]
type = "filesystem"
path = "/vdirsyncer/contacts/"
fileext = ".vcf"

[storage wdconold]
type = "carddav"
verify_fingerprint = "yYy"
verify = false
url = "https://my-url.com:8443/carddav/"
username = "xXx"
password.fetch = ["prompt", "Password for CardDAV"]
```

And now the error message:
```
/usr/virtualenvs/vdirsyncer/bin/vdirsyncer -vdebug discover wdcontacts
debug: Using 1 maximal workers.
debug: Fetching value for password.fetch with prompt strategy.
Password for CardDAV: 
Discovering collections for pair wdcontacts
debug: PROPFIND https://my-url.com:8443/carddav/
debug: {'User-Agent': 'vdirsyncer', 'Content-Type': 'application/xml; charset=UTF-8', 'Depth': '1'}
debug: 
debug:     <d:propfind xmlns:d="DAV:">
debug:         <d:prop>
debug:             <d:resourcetype />
debug:         </d:prop>
debug:     </d:propfind>
debug:     
debug: Sending request...
debug: 207
debug: {'Date': 'Fri, 27 Jan 2017 11:06:14 GMT', 'Keep-Alive': '300', 'Connection': 'Keep-Alive', 'Server': 'Zarafa', 'Content-Length': '921', 'Content-Type': 'application/xml; charset="utf-8"'}
debug: <?xml version="1.0" encoding="UTF-8"?>
debug: <C:multistatus xmlns:D="urn:ietf:params:xml:ns:caldav" xmlns:C="DAV:">
debug:  <C:response>
debug:   <C:href>/carddav/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <C:resourcetype>
debug:      <C:collection/>
debug:     </C:resourcetype>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 200 OK</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug:  <C:response>
debug:   <C:href>/caldav/xXx/FLDPRFX_2F6273A8B84A4BAFA5FE7C09DC666359/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <C:resourcetype>
debug:      <C:collection/>
debug:      <D:calendar/>
debug:     </C:resourcetype>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 200 OK</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug:  <C:response>
debug:   <C:href>/caldav/xXx/FLDPRFX_CEBE10E12D6B4D0F86A5732D28676F1D/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <C:resourcetype>
debug:      <C:collection/>
debug:      <D:calendar/>
debug:     </C:resourcetype>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 200 OK</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug: </C:multistatus>
debug: Given URL is not a homeset URL
debug: GET https://my-url.com:8443/.well-known/carddav
debug: {'User-Agent': 'vdirsyncer', 'Content-Type': 'application/xml; charset=UTF-8'}
debug: None
debug: Sending request...
debug: 404
debug: {'Connection': 'Keep-Alive', 'Server': 'Zarafa', 'Date': 'Fri, 27 Jan 2017 11:06:14 GMT', 'Content-Length': '0', 'Keep-Alive': '300'}

debug: Server does not support well-known URIs.
debug: PROPFIND https://my-url.com:8443/carddav/
debug: {'User-Agent': 'vdirsyncer', 'Content-Type': 'application/xml; charset=UTF-8', 'Depth': '0'}
debug: 
debug:         <d:propfind xmlns:d="DAV:">
debug:             <d:prop>
debug:                 <d:current-user-principal />
debug:             </d:prop>
debug:         </d:propfind>
debug:         
debug: Sending request...
debug: 207
debug: {'Date': 'Fri, 27 Jan 2017 11:06:14 GMT', 'Keep-Alive': '300', 'Connection': 'Keep-Alive', 'Server': 'Zarafa', 'Content-Length': '346', 'Content-Type': 'application/xml; charset="utf-8"'}
debug: <?xml version="1.0" encoding="UTF-8"?>
debug: <C:multistatus xmlns:C="DAV:">
debug:  <C:response>
debug:   <C:href>/carddav/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <C:current-user-principal>
debug:      <C:href>/caldav/xXx/</C:href>
debug:     </C:current-user-principal>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 200 OK</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug: </C:multistatus>
debug: PROPFIND https://my-url.com:8443/caldav/xXx/
debug: {'User-Agent': 'vdirsyncer', 'Content-Type': 'application/xml; charset=UTF-8', 'Depth': '0'}
debug: 
debug:     <d:propfind xmlns:d="DAV:" xmlns:c="urn:ietf:params:xml:ns:carddav">
debug:         <d:prop>
debug:             <c:addressbook-home-set />
debug:         </d:prop>
debug:     </d:propfind>
debug:     
debug: Sending request...
debug: 207
debug: {'DAV': '1, access-control, calendar-access, calendar-schedule, calendarserver-principal-property-search', 'Date': 'Fri, 27 Jan 2017 11:06:14 GMT', 'Keep-Alive': '300', 'Connection': 'Keep-Alive', 'Server': 'Zarafa', 'Content-Length': '329', 'Content-Type': 'application/xml; charset="utf-8"'}
debug: <?xml version="1.0" encoding="UTF-8"?>
debug: <C:multistatus xmlns:D="urn:ietf:params:xml:ns:carddav" xmlns:C="DAV:">
debug:  <C:response>
debug:   <C:href>/caldav/xXx/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <D:addressbook-home-set/>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 404 Not Found</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug: </C:multistatus>
error: The server returned something vdirsyncer doesn't understand. Error message: InvalidXMLResponse("Couldn't find home-set.",)
error: While this is most likely a serverside problem, the vdirsyncer devs are generally interested in such bugs. Please report it in the issue tracker at https://github.com/pimutils/vdirsyncer/issues
error: 1 out of 1 tasks failed.
```

I've tried different directories in the url (combinations of "carddav", "addressbook", username, ...), automatic collection discovery (with = null), and different vdirsyncer versions (0.14.1, 0.14.2.dev22+nga3e39f9).

**Comment by untitaker** [comment]
If you know the direct URL to your addressbook, you can use `collections = null`. If not even that works, your server is broken beyond hope.

**Comment by whitedwarfsun** [comment]
I do not know the direct URL to the addressbook. I tried to guess ([educatedly](http://ntbab.dyndns.org/apache2-default/seite/carddavproviderCE.html) ) but without luck. Is there a way to find the url (apart from diggin in the code of zarafa) ?

Besides: the server is not broke, at least, it does not seem to be: tzpush via thunderbird and direct includes in android do work.

**Comment by untitaker** [comment]
I just noticed, but the URLs in the first HTTP response are:

* `/caldav/xXx/FLDPRFX_2F6273A8B84A4BAFA5FE7C09DC666359/`
* `/caldav/xXx/FLDPRFX_CEBE10E12D6B4D0F86A5732D28676F1D/`

Those may be addressbooks.

They are reported with a resourcetype of `collection`, but the CardDAV RFC requires [that they are reported with the addressbook type as well](https://tools.ietf.org/html/rfc6352#section-5.2).

So your config would look something like this. I'm not sure how many addressbooks you have, I guess you need to try both URLs:

    [pair wdcontacts]
    a = "wdconloc"
    b = "wdconold"
    collections = null

    [storage wdconloc]
    type = "filesystem"
    path = "/vdirsyncer/contacts/"
    fileext = ".vcf"

    [storage wdconold]
    type = "carddav"
    verify_fingerprint = "yYy"
    verify = false

    # Or the other one
    url = "https://my-url.com:8443/caldav/xXx/FLDPRFX_CEBE10E12D6B4D0F86A5732D28676F1D/"
    username = "xXx"
    password.fetch = ["prompt", "Password for CardDAV"]

**Comment by whitedwarfsun** [comment]
These two URLs are my Contacts and Tasks. I am sure about this: vdirsyncer creates subdirectories with equal hashes when I sync caldav using almost the same settings for the server (type=caldav, fileext=.ics, url=https://my-url.com/caldav).

**Comment by whitedwarfsun** [comment]
Btw: the debug log changes, when I use collections = null and url=https://my-url.com:8443/carddav/xXx/Contacts:

```
debug: Using 1 maximal workers.
debug: Fetching value for password.fetch with prompt strategy.

Password for CardDAV: Discovering collections for pair wdcontacts
debug: PROPFIND https://my-url.com:8443/carddav/xXx/Contacts/
debug: {'Depth': '1', 'Content-Type': 'application/xml; charset=UTF-8', 'User-Agent': 'vdirsyncer'}
debug: 
debug:     <d:propfind xmlns:d="DAV:">
debug:         <d:prop>
debug:             <d:resourcetype />
debug:         </d:prop>
debug:     </d:propfind>
debug:     
debug: Sending request...
debug: 207
debug: {'Date': 'Fri, 27 Jan 2017 15:49:59 GMT', 'Connection': 'Keep-Alive', 'Content-Type': 'application/xml; charset="utf-8"', 'Server': 'Zarafa', 'Content-Length': '325', 'Keep-Alive': '300'}
debug: <?xml version="1.0" encoding="UTF-8"?>
debug: <C:multistatus xmlns:C="DAV:">
debug:  <C:response>
debug:   <C:href>/carddav/xXx/Contacts/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <C:resourcetype>
debug:      <C:collection/>
debug:     </C:resourcetype>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 200 OK</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug: </C:multistatus>
debug: Skipping, not of resource type {urn:ietf:params:xml:ns:carddav}addressbook: <Element '{DAV:}response' at 0x7ff4a7de6c28>
debug: Given URL is not a homeset URL
debug: GET https://my-url.com:8443/.well-known/carddav
debug: {'Content-Type': 'application/xml; charset=UTF-8', 'User-Agent': 'vdirsyncer'}
debug: None
debug: Sending request...
debug: 404
debug: {'Content-Length': '0', 'Date': 'Fri, 27 Jan 2017 15:50:00 GMT', 'Connection': 'Keep-Alive', 'Server': 'Zarafa', 'Keep-Alive': '300'}

debug: Server does not support well-known URIs.
debug: PROPFIND https://my-url.com:8443/carddav/xXx/Contacts/
debug: {'Depth': '0', 'Content-Type': 'application/xml; charset=UTF-8', 'User-Agent': 'vdirsyncer'}
debug: 
debug:         <d:propfind xmlns:d="DAV:">
debug:             <d:prop>
debug:                 <d:current-user-principal />
debug:             </d:prop>
debug:         </d:propfind>
debug:         
debug: Sending request...
debug: 207
debug: {'Date': 'Fri, 27 Jan 2017 15:50:00 GMT', 'Connection': 'Keep-Alive', 'Content-Type': 'application/xml; charset="utf-8"', 'Server': 'Zarafa', 'Content-Length': '366', 'Keep-Alive': '300'}
debug: <?xml version="1.0" encoding="UTF-8"?>
debug: <C:multistatus xmlns:C="DAV:">
debug:  <C:response>
debug:   <C:href>/carddav/xXx/Contacts/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <C:current-user-principal>
debug:      <C:href>/caldav/xXx/</C:href>
debug:     </C:current-user-principal>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 200 OK</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug: </C:multistatus>
debug: PROPFIND https://my-url.com:8443/caldav/xXx/
debug: {'Depth': '0', 'Content-Type': 'application/xml; charset=UTF-8', 'User-Agent': 'vdirsyncer'}
debug: 
debug:     <d:propfind xmlns:d="DAV:" xmlns:c="urn:ietf:params:xml:ns:carddav">
debug:         <d:prop>
debug:             <c:addressbook-home-set />
debug:         </d:prop>
debug:     </d:propfind>
debug:     
debug: Sending request...
debug: 207
error: The server returned something vdirsyncer doesn't understand. Error message: InvalidXMLResponse("Couldn't find home-set.",)
error: While this is most likely a serverside problem, the vdirsyncer devs are generally interested in such bugs. Please report it in the issue tracker at https://github.com/pimutils/vdirsyncer/issues
debug: {'Date': 'Fri, 27 Jan 2017 15:50:00 GMT', 'Connection': 'Keep-Alive', 'Content-Type': 'application/xml; charset="utf-8"', 'Server': 'Zarafa', 'Content-Length': '329', 'DAV': '1, access-control, calendar-access, calendar-schedule, calendarserver-principal-property-search', 'Keep-Alive': '300'}
debug: <?xml version="1.0" encoding="UTF-8"?>
debug: <C:multistatus xmlns:D="urn:ietf:params:xml:ns:carddav" xmlns:C="DAV:">
debug:  <C:response>
debug:   <C:href>/caldav/xXx/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <D:addressbook-home-set/>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 404 Not Found</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug: </C:multistatus>
error: 1 out of 1 tasks failed.
```

Also, there seems to be a difference between this URL and any other (like .../xXx/foobar ). When I do not use carddav/xXx/contacts I get only 404 responses and no 207.

**Comment by untitaker** [comment]
I see. I'll think of a way to disable discovery completely when `collections = null` such that you can properly work around this.

**Comment by untitaker** [comment]
@whitedwarfsun Could you check whether https://github.com/pimutils/vdirsyncer/pull/544 fixes your issue?

**Comment by whitedwarfsun** [comment]
Uh, that was quick :)

I've tried again, with branch issue-543, collections = null, full url (i.e. .../carddav/xXx/Contacts ). Here is the output:

```
debug: Using 1 maximal workers.
debug: Fetching value for password.fetch with prompt strategy.
Password for CardDAV: 
Syncing wdcontacts
debug: Loaded status for wdcontacts
debug: PROPFIND https://my-url.com:8443/carddav/xXx/Contacts/
debug: {'Content-Type': 'application/xml; charset=UTF-8', 'Depth': '1', 'User-Agent': 'vdirsyncer'}
debug: <?xml version=1.0 encoding=utf-8 ?>
debug:             <D:propfind xmlns:D=DAV:>
debug:                 <D:prop>
debug:                     <D:resourcetype/>
debug:                     <D:getcontenttype/>
debug:                     <D:getetag/>
debug:                 </D:prop>
debug:             </D:propfind>
debug:             
debug: Sending request...
debug: 207
debug: {'Keep-Alive': '300', 'Date': 'Sat, 28 Jan 2017 19:35:30 GMT', 'Connection': 'Keep-Alive', 'Server': 'Zarafa', 'Content-Type': 'application/xml; charset=utf-8', 'Content-Length': '418'}
debug: <?xml version=1.0 encoding=UTF-8?>
debug: <C:multistatus xmlns:C=DAV:>
debug:  <C:response>
debug:   <C:href>/carddav/xXx/Contacts/</C:href>
debug:   <C:propstat>
debug:    <C:prop>
debug:     <C:resourcetype>
debug:      <C:collection/>
debug:     </C:resourcetype>
debug:     <C:getcontenttype>text/calendar</C:getcontenttype>
debug:     <C:getetag>1323650342</C:getetag>
debug:    </C:prop>
debug:    <C:status>HTTP/1.1 200 OK</C:status>
debug:   </C:propstat>
debug:  </C:response>
debug: </C:multistatus>
debug: Already normalized: '/carddav/xXx/Contacts/'
debug: Skipping '/carddav/xXx/Contacts/', is collection.
```

Unfortunately, it isn't working.

I've tried collections = ["contacts"] (and removed the contacts part in the url) but again without luck.

**Comment by untitaker** [comment]
Okay, at this point I can't do much anymore. I pushed a new commit to that branch that you could try out, but that really should not be necessary.

**Comment by untitaker** [comment]
I pushed the obviously good changes to master, but I don't know what to do about your other issue. I suspect there is in fact a bug in your server software. I'd like to see a HTTP traffic log from a client that works (particularly from when you set it up)

**Comment by untitaker** [comment]
@whitedwarfsun any update on this?

**Comment by untitaker** [comment]
Closing due to inactivity.

**Comment by whitedwarfsun** [comment]
Sorry for the lag. However, I want to make a final comment on this issue.

I tried again, with your various commits, with different URLs, with various settings for collection, but without luck.
Then I looked further into the setup and I think the problem is that in my install of Zarafa there is no carddav server running. For this, a Zarafa plugin like Sabre (or similar) would be required but Sabre is not installed. Yet, I do have access to contacts on my server, but they seem to be served by z-push via ActiveSync. Hence the confusion. Sorry for that and thank you for your fast help. The issues can be kept closed.
